lib/Dialect/LWE/Conversions/LWEToOpenfhe/LWEToOpenfhe.cpp

```diff
@@ -22,125 +30,286 @@
+      op.emitError() << "HEIR does not yet support coefficient encoding "
+                        " when targeting OpenFHE";
+      return failure();
+    }
+    if (isa<lwe::PolynomialEvaluationEncodingAttr>(op.getEncoding()) ||
+        isa<lwe::FullCRTPackingEncodingAttr>(op.getEncoding())) {
+      rewriter.replaceOpWithNewOp<openfhe::MakePackedPlaintextOp>(
+          op, plaintextType, cryptoContext, input);
+
+      return success();
+    }
+    // else: encoding isn't support explicitly:
+    op.emitError(
+        "Unexpected encoding while targeting OpenFHE. "
+        "If you expect this type of encoding to be supported "
+        "for the OpenFHE backend, please file a bug report.");
+    return failure();
   }
+};
 
-  return success();
-}
+struct LWEToOpenfhe : public impl::LWEToOpenfheBase<LWEToOpenfhe> {
+  void runOnOperation() override {
+    MLIRContext *context = &getContext();
+    auto *module = getOperation();
+    ToOpenfheTypeConverter typeConverter(context);
+
+    ConversionTarget target(*context);
+    target.addLegalDialect<openfhe::OpenfheDialect>();
+    target.addIllegalDialect<bgv::BGVDialect>();
+    target.addIllegalDialect<ckks::CKKSDialect>();
+    target.addIllegalDialect<lwe::LWEDialect>();
+    // We can keep the following ops, which the emitter can handle directly
+    target.addLegalOp<lwe::ReinterpretUnderlyingTypeOp, lwe::RLWEDecodeOp>();
+
+    RewritePatternSet patterns(context);
+    addStructuralConversionPatterns(typeConverter, patterns, target);
+
+    target.addDynamicallyLegalOp<func::FuncOp>([&](func::FuncOp op) {
+      bool hasCryptoContextArg = op.getFunctionType().getNumInputs() > 0 &&
+                                 mlir::isa<openfhe::CryptoContextType>(
+                                     *op.getFunctionType().getInputs().begin());
+      return typeConverter.isSignatureLegal(op.getFunctionType()) &&
+             typeConverter.isLegal(&op.getBody()) &&
+             (!containsDialects<lwe::LWEDialect, bgv::BGVDialect,
+                                ckks::CKKSDialect>(op) ||
+              hasCryptoContextArg);
+    });
+
+    patterns.add<
+        /////////////////////
+        // LWE Op Patterns //
+        /////////////////////
+
+        // Update Func Op Signature
+        AddCryptoContextArg,
+
+        // Handle LWE encode and en/decrypt
+        // Note: `lwe.decode` is handled directly by the OpenFHE emitter
+        ConvertEncodeOp, ConvertEncryptOp, ConvertDecryptOp,
+
+        // Scheme-agnostic RLWE Arithmetic Ops:
+        ConvertBinOp<lwe::RAddOp, openfhe::AddOp>,
+        ConvertBinOp<lwe::RSubOp, openfhe::SubOp>,
+        ConvertBinOp<lwe::RMulOp, openfhe::MulNoRelinOp>,
+        ConvertUnaryOp<lwe::RNegateOp, openfhe::NegateOp>,
+
+        ///////////////////////////////////
+        // Scheme-Specific Op Patterns   //
+        ///////////////////////////////////
+
+        // AddPlain (//FIXME: Not really all that scheme-specific)
```

Comment:
Here and below, remove fixmes or resolve them.

You want an `lwe.add_plain/sub_plain/mul_plain` op?

Reply:
I think eventually it'd be nicer to just support ptxt types on the existing lwe ops, but I'll leave that as a TODO for #1193